ssl/rio/rio_notifier.c

```diff
@@ -0,0 +1,247 @@
+#include "internal/rio_notifier.h"
+#include <openssl/bio.h>
+
+/*
+ * Sets a socket as close-on-exec, except that this is a no-op if we are certain
+ * we do not need to do this or the OS does not support the concept.
+ */
+static int set_cloexec(int fd)
+{
+#if !defined(SOCK_CLOEXEC) && defined(FD_CLOEXEC)
+    return fcntl(fd, F_SETFD, FD_CLOEXEC) >= 0;
+#else
+    return 1;
+#endif
+}
+
+#if RIO_NOTIFIER_METHOD == RIO_NOTIFIER_METHOD_SOCKET
+
+/* Create a close-on-exec socket. */
+static int create_socket(int domain, int socktype, int protocol)
+{
+    int fd;
+
+#if defined(OPENSSL_SYS_WINDOWS)
+    static const int on = 1;
+
+    /*
+     * Use WSASocketA to create a socket which is immediately marked as
+     * non-inheritable, avoiding race conditions if another thread is about to
+     * call CreateProcess.
+     */
+    fd = WSASocketA(domain, socktype, protocol, NULL, 0,
+                    WSA_FLAG_NO_HANDLE_INHERIT);
+    if (fd < 0)
+        return -1;
+
+    /* Prevent interference with the socket from other processes on Windows. */
+    if (setsockopt(lfd, SOL_SOCKET, SO_EXCLUSIVEADDRUSE, &on, sizeof(on)) < 0) {
+        BIO_closesocket(fd);
+        return -1;
+    }
+
+#else
+# if defined(SOCK_CLOEXEC)
+    socktype |= SOCK_CLOEXEC;
+# endif
+
+    fd = BIO_socket(domain, socktype, protocol, 0);
+    if (fd < 0)
+        return -1;
+
+    /*
+     * Make socket close-on-exec unless this was already done above at socket
+     * creation time.
+     */
+    if (!set_cloexec(fd)) {
```

Comment:
I'm not sure I understand this comment.  The socket is created by this function, so this is socket creation time, no?

Reply:
The socket is created with the `BIO_socket()` call on new line 49.